### DISCUSSION TITLE
Packet length error, ESP32 + RS485 <--> ESP32 +RS485

### DISCUSSION THREAD

**Discussion by jax-gin** [eModbus/eModbus#378] [Q&A] [answered]
Hello, I am using two ESP32s with the RS485 module to communicate between them, but when it sends the request I get the following result:

__ OK __
[D] 31| ModbusServer.cpp     [  16] registerWorker: Registered worker for 01/03
[V] 32| RTUutils.cpp         [ 112] calculateInterval: Calc interval(115201)=1750
[D] 42| ModbusServerRTU.cpp  [  99] doBegin: Server task 1073451972 started. Interval=1750
[V] 44| RTUutils.cpp         [ 262] receive: S/1773us without data after 1
[V] 55| RTUutils.cpp         [ 272] receive: S/[V] Raw buffer received: @3FFB952C/1:
  | 0000: 00                                                |.               |
[D] 66| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB973C/1:
  | 0000: E5                                                |.               |
[E] 86| ModbusServerRTU.cpp  [ 259] serve: RTU receive: E5 - Packet length error
[D] 2088| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB9960/1:
  | 0000: E0                                                |.               |
[D] 4100| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB9974/1:
  | 0000: E0                                                |.               |
[D] 6112| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB9988/1:
  | 0000: E0                                                |.               |
[D] 8124| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB999C/1:
  | 0000: E0                                                |.               |
[V] 9634| RTUutils.cpp         [ 262] receive: S/1760us without data after 9
[V] 9634| RTUutils.cpp         [ 272] receive: S/[V] Raw buffer received: @3FFB952C/9:
  | 0000: 01 03 00 00 00 04 44 09  00                       |......D..       |
[D] 9645| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB99C4/7:
  | 0000: 01 03 00 00 00 04 44                              |......D         |
[D] 9665| ModbusServerRTU.cpp  [ 178] serve: Request received.
[D] 9666| ModbusServer.cpp     [  22] getWorker: Need worker for 01-03 : Worker found for 01/03
[D] 9676| ModbusServerRTU.cpp  [ 201] serve: Callback found.
[D] 9686| ModbusServerRTU.cpp  [ 208] serve: Callback called.
[V] Callback response: @3FFB99F4/3:
  | 0000: 01 83 02                                          |...             |
[D] Sent packet: @3FFB9A08/3:
  | 0000: 01 83 02                                          |...             |
[D] 9708| ModbusServerRTU.cpp  [ 245] serve: Response sent.
[V] 9711| RTUutils.cpp         [ 262] receive: S/1763us without data after 1
[V] 9721| RTUutils.cpp         [ 272] receive: S/[V] Raw buffer received: @3FFB952C/1:
  | 0000: 00                                                |.               |
[D] 9731| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB9A08/1:
  | 0000: E5                                                |.               |

And in the side of the client i have:

__ OK __
[D] 47| ModbusClientRTU.cpp  [ 117] setTimeout: Timeout set to 2000
[V] 48| RTUutils.cpp         [ 112] calculateInterval: Calc interval(115201)=1750
[D] 59| ModbusClientRTU.cpp  [  91] doBegin: Client task 1073451860 started. Interval=1750
[V] 10060| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 10060| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
[D] 10070| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 10070| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[D] 10071| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB959C/6:
  | 0000: 01 03 00 00 00 04                                 |......          |
[D] 10093| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[V] 10095| RTUutils.cpp         [ 262] receive: C/1770us without data after 1
[V] 10106| RTUutils.cpp         [ 272] receive: C/[V] Raw buffer received: @3FFB95F8/1:
  | 0000: 00                                                |.               |
[D] 10116| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB959C/1:
  | 0000: E5                                                |.               |
[D] 10137| ModbusClientRTU.cpp  [ 282] handleConnection: Error response (1 bytes) received.
[V] Data: @3FFB959C/1:
  | 0000: E5                                                |.               |
[D] 10148| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FFB959C/3:
  | 0000: 01 83 E5                                          |...             |
[E] 10168| main_1.cpp           [  57] handleError: Error response: E5 - Packet length error

I don't know why i have a extra byte in the request received in the server.

**Comment by Miq1** [comment, marked as answer]
That extra byte most probably is generated by your RS485 adapters. There are adapters that will go floating while switching from read to write mode and back. This open drain is registered at the ESP's UART as a 0x00 byte.

This is quite common, you will find a pinned discussion about the reasons.

There are some remedies:
- get other RS485 adapters that have a defined state while switching. This is hard to determine, every vendor may have some or not - you will only know if you try. RS485 adapters with auto half duplex switching will not have this issue, I would recommend these in any case.
- try to add pulldown resistors to the RX line to the ESP - 10K or 4K7 may do.
- Use the ``void skipLeading0x00(bool onOff = true);`` call to have the 0x00 skipped if it comes in front of a message. This may work as long as the time between the artificial 0x00 and the message proper is below the RTU quiet time, so it gets added to the message.

**Reply by jax-gin** [reply]
where a have to use the `void skipLeading0x00(bool onOff = true);` ?? in the server setup() or in the client setup()?

**Reply by Miq1** [reply]
If you are using the same adapters for both, on both.

**Reply by jax-gin** [reply]
nothing works, but here's my code, maybe I made a mistake when adding that function


CLIENT
```
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Set up Serial2 connected to Modbus RTU
  RTUutils::prepareHardwareSerial(Serial2);
  Serial2.begin(BAUDRATE, SERIAL_8N1, RXPIN, TXPIN);

void skipLeading0x00(bool onOff = true);

// Set up ModbusRTU client.
// - provide onData handler function
  MB.onDataHandler(&handleData);
// - provide onError handler function
  MB.onErrorHandler(&handleError);
// Set message timeout to 2000ms
  MB.setTimeout(2000);
// Start ModbusRTU background task
  MB.begin(Serial2, 1);
}

```
SERVER

```
void setup() {
// Init Serial monitor
  Serial.begin(115200);
  while (!Serial) {}
  Serial.println("__ OK __");

// Init Serial2 connected to the RTU Modbus
// (Fill in your data here!)
  RTUutils::prepareHardwareSerial(Serial2);
  Serial2.begin(115200, SERIAL_8N1, 16, 17);

// Register served function code worker for server 1, FC 0x03
  void skipLeading0x00(bool onOff = true);
  MBserver.registerWorker(0x01, READ_HOLD_REGISTER, &FC03);

// Start ModbusRTU background task
  MBserver.begin(Serial2);
}
```

**Reply by Miq1** [reply]
The call will be ``MB.skipLeading0x00();`` and ``MBserver.skipLeading0x00()`` - it will set an option for the objects.

**Reply by jax-gin** [reply]
the communication works but the 0x00 byte are still there: 

CLIENT 

[V] 2713563| ModbusMessage.cpp    [ 477] checkData: Check data #3
[D] 2713563| ModbusClientRTU.cpp  [ 161] addRequestM: request for 01/03
[D] 2713573| ModbusClientRTU.cpp  [ 244] addToQueue: RC=01
[D] 2713573| ModbusClientRTU.cpp  [ 172] addRequestM: RC=00
[D] 2713574| ModbusClientRTU.cpp  [ 262] handleConnection: Pulled request from queue
[D] Sent packet: @3FFB97FC/6:
  | 0000: 01 03 00 01 00 04                                 |......          |
[D] 2713595| ModbusClientRTU.cpp  [ 267] handleConnection: Request sent.
[V] 2713652| RTUutils.cpp         [ 262] receive: C/1752us without data after 14
[V] 2713652| RTUutils.cpp         [ 272] receive: C/[V] Raw buffer received: @3FFB36FC/14:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04 0D 14 00        |..............  |
[D] 2713673| RTUutils.cpp         [ 437] receive: C/[D] Received packet: @3FFB97FC/12:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04 0D              |............    |
[D] 2713684| ModbusClientRTU.cpp  [ 282] handleConnection: Data response (12 bytes) received.
[V] Data: @3FFB97FC/12:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04 0D              |............    |
[D] 2713704| ModbusClientRTU.cpp  [ 303] handleConnection: Response generated.
[V] Response packet: @3FFB97FC/12:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04 0D              |............    |
[V] got float: @3FFC1D58/4:
  | 0000: 02 00 01 00                                       |....            |
[V] got float swapped: @3FFC1D58/4:
  | 0000: 02 00 01 00                                       |....            |
[V] got float: @3FFC1D5C/4:
  | 0000: 04 00 03 00                                       |....            |
[V] got float swapped: @3FFC1D5C/4:
  | 0000: 04 00 03 00                                       |....            |
Requested at 2713.563s:
   0001:    0.000
   0003:    0.000

SERVER

[V] 2639897| RTUutils.cpp         [ 272] receive: S/[V] Raw buffer received: @3FFB95E4/9:
  | 0000: 01 03 00 01 00 04 15 C9  00                       |.........       |
[D] 2639907| RTUutils.cpp         [ 437] receive: S/[D] Received packet: @3FFB9574/7:
  | 0000: 01 03 00 01 00 04 15                              |.......         |
[D] 2639928| ModbusServerRTU.cpp  [ 178] serve: Request received.
[D] 2639928| ModbusServer.cpp     [  22] getWorker: Need worker for 01-03 : Worker found for 01/03
[D] 2639938| ModbusServerRTU.cpp  [ 201] serve: Callback found.
[D] 2639949| ModbusServerRTU.cpp  [ 208] serve: Callback called.
[V] Callback response: @3FFB95A4/11:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04                 |...........     |
[D] Sent packet: @3FFB9548/11:
  | 0000: 01 03 08 00 01 00 02 00  03 00 04                 |...........     |
[D] 2639971| ModbusServerRTU.cpp  [ 245] serve: Response sent.
[V] 2640920| RTUutils.cpp         [ 262] receive: S/1752us without data after 9

**Reply by Miq1** [reply]
> the communication works but the 0x00 byte are still there:

Of course. The function now only is ignoring it - there is no way to modify the hardware that is causing it.

**Reply by jax-gin** [reply]
Okay, but if I replace one of the ESP32 with another device, it will read the extra byte, and the communication will fail, right?

The only way for this to work is by using RS485 adapters with automatic half-duplex switching.

The 10k pulldown resistor doesn't work either.

**Reply by Miq1** [reply]
Are you sure you have the wiring right? Biasing resistors at one end of the bus?

But yes, my choice would be other adapters.

**Reply by jax-gin** [reply]
here's the wiring: 

![image](https://github.com/user-attachments/assets/b6bf09f3-475f-47df-a239-61f5549112c7)

![image](https://github.com/user-attachments/assets/029b0fce-c4de-41b1-9020-4b68bd132abf)

![image](https://github.com/user-attachments/assets/74dd3b23-c298-4190-abda-3c3259b9f010)

Another question, where i can find the discussion before mentioned?

**Comment by Miq1** [comment]
The discussion is [here](https://github.com/eModbus/eModbus/discussions/110).

Bert found a pull_up_ resistor of 470 Ohms only on the RX line solved the issue. Sorry, I had that in mind differently, it is pull up instead and 4k7 is much too weak.

**Reply by jax-gin** [reply]
The pull-up resistor solved the client 0x00 byte issue, but not the server 0x00 byte. So the request is received correctly, but the client still needs the function `skipLeading0x00(true);`.

**Reply by Miq1** [reply]
That is strange. Both should behave indentically. ❓

**Reply by jax-gin** [reply]
Well, I'm waiting for a new adapter. The pull-up resistors are not working today, so I gave up on that.
